tpm2/src/util/tpm2_kmyth_seal.c

```diff
@@ -30,12 +30,443 @@ extern const cipher_t cipher_list[];
+
+  // validate non-empty plaintext buffer specified
+  if (input_length == 0 || input == NULL)
+  {
+    kmyth_log(LOG_ERR, "no input data ... exiting");
+    return 1;
+  }
+
+  // encrypt (wrap) input data read in (e.g., client certificate private .pem)
+  if (kmyth_encrypt_data(input, input_length,
+                         cipher, &wrapped_data, &wrapped_data_size,
+                         &wrapKey, &wrapKey_size))
+  {
+    kmyth_log(LOG_ERR, "unable to encrypt (wrap) data ... exiting");
+    return 1;
+  }
+
+  kmyth_log(LOG_DEBUG, "input data wrapped");
+
+  // private and public blobs for sealed data object
+  TPM2B_PRIVATE sealedData_private;
+  TPM2B_PUBLIC sealedData_public;
+
+  // init private and public blobs to empty
+  sealedData_private.size = 0;
+  sealedData_public.size = 0;
+
+  // Seal the wrapping key to the TPM using the Storage Key (SK)
+  if (tpm2_kmyth_seal_data(sapi_ctx,
+                           wrapKey,
+                           wrapKey_size,
+                           storageKey_handle,
+                           objAuthVal,
+                           objPcrList,
+                           objAuthVal,
+                           objPcrList,
+                           objAuthPolicy,
+                           &sealedData_public, &sealedData_private))
+  {
+    kmyth_log(LOG_ERR, "unable to seal data ... exiting");
+    kmyth_clear_and_free(wrapKey, wrapKey_size);
+    free(wrapped_data);
+    kmyth_clear(objAuthVal.buffer, objAuthVal.size);
+    tpm2_free_resources(&sapi_ctx);
+    return 1;
+  }
+
+  // Clean-up:
+  //   - done with unencrypted wrapping key (now have sealed version)
+  //   - done with authVal
+  kmyth_clear_and_free(wrapKey, wrapKey_size);
+  kmyth_clear(objAuthVal.buffer, objAuthVal.size);
+
+  if (tpm2_kmyth_create_ski_string(output,
+                                   output_length,
+                                   NULL,
+                                   0,
+                                   objPcrList,
+                                   storageKey_public,
+                                   storageKey_private,
+                                   cipher.cipher_name,
+                                   sealedData_public, sealedData_private,
+                                   wrapped_data, wrapped_data_size))
+  {
+    kmyth_log(LOG_ERR, "error writing data to .ski format ... exiting");
+    free(wrapped_data);
+    tpm2_free_resources(&sapi_ctx);
+    return 1;
+  }
+
+  // done, so free any allocated resources that remain
+  free(wrapped_data);
+  tpm2_free_resources(&sapi_ctx);
+
+  return 0;
+}
+
+//############################################################################
+// tpm2_kmyth_unseal()
+//############################################################################
+int tpm2_kmyth_unseal(uint8_t * input, size_t input_len,
+                      uint8_t ** output, size_t *output_len,
+                      char *auth_string, char *owner_auth_passwd)
```

Comment:
I'd really like to shift passwords to `uint8_t*`s with tracked lengths so it's possible to use key material as a password.